qrenderdoc: Allow vkCmdExecuteCommands events to be viewed in the API inspector

aejsmith opened the pull request:
Currently, selecting a vkCmdExecuteCommands event in the event browser will cause the API inspector window to show the final child event of the command buffer being executed (i.e. the "vkEndCommandBuffer" marker rather than the vkCmdExecuteCommands event itself), which means it is not possible to inspect that API call and calls just prior to it in the API inspector (e.g. I have some vkCmdPipelineBarrier calls prior to vkCmdExecuteCommands that I want to be able to inspect).

This happens because EventBrowser::AddDrawcalls sets COL_SELECT_ID for events with children to the event ID of the last child event. This behaviour makes sense for debug markers because they're not real events (I think is the only other case where an event can have children?).

Therefore this change makes it possible to inspect vkCmdExecuteCommands events by setting COL_SELECT_ID to the ID of the event itself rather than that of the last child if the event has eDraw_CmdList set. This keeps the previous behaviour for other types of event with children.

This seemed like the best way to handle this to me, though if you can think of anything better let me know and I can change this.

baldurk commented:
I'm not sure this is the right fix - I think the problem is more complex. The current behaviour is the way it is so that when you select a node with children (any node, whether it be a debug marker or a command list execute), then the current state reflects the state of the frame _after_ it has run. This is much more consistent and easy to understand than if it reflects the state _before_ it has run, since that means you then need to expand and pick out the last event to see the results.

But equally like you say you want to be able to see the API events for the actual root event itself, not for the last event. This would be the same if you had some barriers before a marker region, they'd be lost.

I think the fix might have to instead be in the API inspector to be able to distinguish between the 'current event' (the last child) and the 'selected event' (which node you actually have selected). So it would mean the API inspector would show different events depending on whether you selected a parent node or its last child, even though otherwise to the UI it's identical. That would require solving a couple of issues e.g. if you click back and forth there won't be any callbacks firing but the selected node would change.

Side-note: bear in mind, qrenderdoc is not really used by anyone yet as it's still work in progress. I don't mind the PR at all, just wanted to clarify that few people will see any fix that only applies there.

aejsmith commented:
Just to make sure I've understood right, you're saying that the API inspector should view the state at the selected event, whereas everything else should continue to see the state at the last child? That sounds reasonable. I can have a look at doing that if I get some time.

Yeah, I'm aware qrenderdoc is still work-in-progress, but I'm using it because as far as I know it's currently the best Vulkan debugger available on Linux. It's already pretty useful :)

baldurk commented:
I'm glad qrenderdoc is useful, I did try and get it to "minimum viable debugger" before I went afk in December :smile:.

And yeh you understand exactly right - for example in the texture viewer, if you select a `vkCmdExecuteCommands` then you want to be able to see the things rendered in it without having to go any further. This means that if you have a whole bunch of executes one after the other, you can move up and down them and see the result of the one you have selected. This happens because the current eventID goes to the last child.

The API inspector will need to be a special case that should know if you have the marker selected or the last child selected, even though to the rest of the UI it will look like it's just the last child. I'd suggest you should add to `ILogViewerForm` a new callback for `OnEventHighlighted` which would fire right before `OnEventSelected` does currently, but it would show the event highlighted. Then if in the event browser you click from a last-child to a marker, `OnEventSelected` doesn't fire but `OnEventHighlighted` would. Likewise you'd be able to query out the highlighted event as supposed to the 'current' event in the `CaptureContext`.

Then the API inspector can listen and look up the events for the highlighted event rather than the selected event. Possibly in theory, some other window will want to make that distinction too.

The terminology there isn't great so maybe you have a better idea for naming, or another way to propagate that information around?

I'll leave this PR open in case you want to update it, or if you'll make a new one then feel free to close.

aejsmith commented:
Just pushed a new version. The terminology I went with was "current event" (which would be the last child for an event with children) and "selected event". I renamed the current OnEventSelected callback to OnEventChanged (for when the current event changed) and added OnSelectedEventChanged for when the selected event changed. The API inspector now uses the selected event, while everything else still uses the current event.

As far as I can see this is behaving right - selecting a vkCmdExecuteCommands shows that in the API inspector while everything else is showing the state after it's completed.

baldurk commented on qrenderdoc/Code/CaptureContext.h:
```diff
@@ -42,7 +42,17 @@ struct ILogViewerForm
+  // the API inspector uses the selected event ID to display the API calls of
+  // that event rather than of the last child.
+  virtual void OnSelectedEventChanged(uint32_t eventID) {}
+  virtual void OnEventChanged(uint32_t eventID) {}
```
I'd rather have these be pure virtual to keep the ILogViewerForm as a pure interface, even if it means adding empty implementations in the log viewers.

aejsmith replied:
Sure, I'll change it.

baldurk commented on qrenderdoc/Windows/EventBrowser.cpp:
```diff
@@ -103,25 +103,25 @@ void EventBrowser::OnLogfileLoaded()
   ui->events->expandItem(frame);
 
-  m_Ctx->SetEventID(this, lastEID);
+  m_Ctx->SetEventID(this, 0, lastEID);
```
This is a bit tricky. I'm not sure we can pass in 0, as 0 isn't a valid eventID so things that rely on having a valid eventID might break down (possibly nothing does at the moment).

On the other hand we have the frame node selected which is kind of outwith the list of events. Perhaps we should pass lastEID for both here, since the last even should be a `Present()` type marker? Does that cause any problems with the event browser behaviour?

aejsmith replied:
The API inspector handles having a null draw call so that being 0 doesn't cause a problem, but yes, I think it probably makes more sense to use lastEID there (so the focus is entirely on the last event of the frame to begin with).

aejsmith commented:
Pushed a new version with those changes.
